### DISCUSSION TITLE
IBM Open Z Editor v4 triggers a TextDocumentChangeEvent while working with copybooks

### DISCUSSION THREAD

**Discussion by curiousDev288** [IBM/zopeneditor-about#398] [General]
Hi,

I was able to configure copybooks inside a cobol program by using the zapp.json and it was working untill v3.4.0. With the v4.0, everytime i update the path of the copybooks inside the zapp.json, it is triggering a textDocumentChange event within the cobol program. Is there any specific reason this event is being emitted?

Thanks

**Comment by sdaimwood** [comment]
Hi @curiousDev288, those events are being sent now as an enhancement to the way Z Open Editor handles changes in ZAPP files. Previously, the entire extension had to restart when a change was made in your ZAPP file. Now, when a ZAPP change is detected, an event is sent to the already-running language servers to let them know that the search paths have changed. 

Everything should still be working as expected and much smoother than it was previously with the extension restarting with every change. Are you experiencing any issues with loading copybooks?

**Reply by curiousDev288** [reply]
The copybook processing is working seamlessly, but there is a scenario where the current opened cobol program is opened in browse mode, and the trigger of copybook download is updating the cobol program and marking it as dirty. I was wondering if some identifier was being sent from the open z editor event so that I could distinguish the incoming change event and make sure to restrict the current opned cobol programs state to be altered.

**Reply by benjamin-santos** [reply]
Hi @curiousDev288. The Cobol program should not be marked dirty after modifying the ZAPP file and downloading copybooks. Would you mind sharing steps to recreate the issue so we can take a look? Please include what editors are visible when you modify the ZAPP File.
Thanks.

**Reply by curiousDev288** [reply]
Hi @benjamin-santos Please find the steps which we are following -
1. Open XYZ.cbl file(contains a few copybook programs)
2. Our program logic identifies the copybooks in the program.
   3. The zapp.yaml is updated with the path of the downloaded copybook programs.
4. As soon as zapp.yaml is updated, it is triggering a textDocumentChangeEvent event with contentChange object,
	contentChange {
		rangeLength: 1,
		offsetRange: 0,
		textContent: '0'
	}
5. This change event is updating the current opened cobol program in the editor. 

Please do let me know, if you need any further details.

Thanks.

**Reply by curiousDev288** [reply]
Hi @benjamin-santos, did you find anything related to the issue?

**Reply by benjamin-santos** [reply]
Hello @curiousDev288 
The textDocumentChangeEvent events are expected - as Saile mentioned when the ZAPP file is modified, open programs are now re-parsed to incorporate the new ZAPP changes. The textDocumentChangeEvent is a part of this reparse for visible programs.

That being said, these programs should not be marked as dirty and the reparse should not affect files that are in read-only mode. If you could provide a screen recording of this, we can take a closer look.

I am curious about one step in the scenario you described: `3. The zapp.yaml is updated with the path of the downloaded copybook programs.`. If you have defined a Cobol property group in your ZAPP file that uses remote libraries, Z Open Editor will download copybooks from that library to a temporary folder that is automatically used when resolving copy statements. So you should not have to update your ZAPP file to point to this temporary folder. If it is a different use case, please provide some more details for this as well. 

Thank you.

**Comment by curiousDev288** [comment]
Hi @benjamin-santos 
I have tried the below steps to create a poc for reproducing the issue - 
1. The cobol file is a read-only file.
2. I have a zapp.yaml file which does not location entries attached to it.
3. I have a python script, which once triggered will update the zapp.yaml file and add entries to location key.
4. If the cobol file is in focus, you would observe in the shared video that the file is now dirty. And also there is an error thrown from vscode which displays there was save failed for the read only file.

Please find the attached video below - 
https://github.com/IBM/zopeneditor-about/assets/100760141/8386ed39-2632-445b-a035-69970c2e0524

**Reply by curiousDev288** [reply]
Hi @benjamin-santos, did you get a chance to look at the shared recording? Please do let me know if you need any further details.

**Reply by curiousDev288** [reply]
Hi @benjamin-santos, @sdaimwood can we get a fix for this issue?

**Reply by benjamin-santos** [reply]
Hi @curiousDev288. Thank you for attaching a video of your issue. I was able to recreate the issue you are experiencing and I know what is causing it. I have created an issue in our backlog for this.

As you noted in Step 4., if the Cobol file is in focus when you run the script the file will become dirty. A workaround you can use until we have a fix ready for this issue is running your script without the Cobol program in focus. If it is not visible in your text editor, it will not be marked dirty when the script is run.

**Reply by curiousDev288** [reply]
Hi @benjamin-santos, thanks for adding the issue in backlog. Appreciate your help.

Is there any possible way we could track the backlog ticket so that we get an idea on the solution timeline.

**Reply by sdaimwood** [reply]
HI @curiousDev288, I've created an issue for this here: https://github.com/IBM/zopeneditor-about/issues/409

We'll update and close that item when we release a fix. Thank you for the help finding and reporting the bug.

**Reply by BenoitEbner** [reply]
Hi @sdaimwood , do you have any idea on the resolution time?
We also have this problem in my company.

thanks!
